packages/opentelemetry-core/src/trace/TraceState.ts

```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as types from '@opentelemetry/types';
+
+// TODO validate maximum number of items
+// const maximumTraceStateItems = 32;
+
+const maximumTraceStateLength = 512;
+export interface InternalTraceState {
+  [key: string]: string;
+}
+
+export class TraceState implements types.TraceState {
+  private internalState: InternalTraceState = {};
+
+  constructor(internalState?: InternalTraceState) {
+    if (internalState) this.internalState = internalState;
+  }
+
+  get(name: string): string | undefined {
+    return this.internalState[name];
+  }
+
+  keys(): string[] {
+    return Object.keys(this.internalState);
+  }
+
+  set(name: string, value: string): void {
+    this.internalState = {
+      // ensure that the new key ends up in the beginning of the list
```

Comment:
JavaScript has no guarantee on key order and the insertion order doesn't change this fact.

Reply:
Another option is to use `list` to keep all the states in order. But that makes `set` operation costly, as we have to iterate on the list to find (and remove) existing duplicate keys/entries.

Reply:
This adds code and space complexity but here's an idea: you can look to an LRU type datastructure, where you have a linked list<items> and a map<key, item> . You maintain ordering through the linked list but get immediate look ups through the map. When updating an existing item, you get the item from the map and bump it to the front of the linked list.

Reply:
> Another option is to use list to keep all the states in order.

IMO this won't add big overhead, given that there can be a maxi. of 32 entries/items in a list.

Reply:
A `Map` would preserve ordering. It's however slower than an object for simple operations. I like the idea above that basically replicates the behavior of a `Map` in a potentially faster way. I'm fine either way, but eventually we should benchmark the different approaches and go with the faster one.

Reply:
One or the other should be done right away, otherwise the order is not guaranteed. I feel this shouldn't be a TODO. Let's use a `Map` for now as it's the simplest approach.

Reply:
Done in https://github.com/open-telemetry/opentelemetry-js/pull/57/commits/c86a044efb589975b85e30f973f36b28ecb5090a